### PR TITLE
export CSV backend

### PR DESCRIPTION
**Pull request by jayellul**

### CODE REVIEW

File: .env.sample [hunk 1]
```diff
@@ -1,3 +0,0 @@
-REDIS_URL=redis://localhost:6379
```
Comment (EvilKanoa) [inline .env.sample]: Don't delete this file. It's useful when you first clone the repo so you can setup your personal `.env` based on it.

File: out.csv [hunk 2]
```diff
@@ -0,0 +1 @@
+_id,firstName,lastName,sector,salary,taxableBenefits,employer,title,province,year,positionClass,severance,original,
```
Comment (EvilKanoa) [inline out.csv]: I don't think this file is actually used, does it need to be committed?

File: src/routes/sunshine.js [hunk 3]
```diff
@@ -1,3 +1,4 @@
+
```
Comment (EvilKanoa) [inline src/routes/sunshine.js]: Don't need a newline at the start of the file.

File: src/routes/sunshine.js [hunk 4]
```diff
@@ -94,8 +95,54 @@ module.exports = (app, db) => {
         res.status(200).json(db.getFields());
     };
 
+    const exportHandler = async (req, res) => {
+        const query = queryFromReq(req);
+
+        // headers - array of strings
+        const headersToCSV = (headers) => {
```
Comment (EvilKanoa) [inline src/routes/sunshine.js:101]: If you make use of the array functions `map` and `join` you can do this whole function in one line, or even just remove the function and do it directly below.
```JavaScript
const headersToCSV = (headers) => `${headers.map((s) => `"${s}"`).join(',')}\n`;
```

File: src/routes/sunshine.js [hunk 5]
```diff
@@ -94,8 +95,54 @@ module.exports = (app, db) => {
...
+                }
+            }
+            line += '\n';
+            return line;
+        }
+
+        try {
+            const cursor = await db.getCursor(query);
+            const csvHeaders = db.getFields();
+            res.setHeader('Content-disposition', 'attachment; filename=SundialExport.csv');
```
Comment (EvilKanoa) [inline src/routes/sunshine.js:129]: I don't know if these response headers are correct, but it does work with them so they work. Just want to make sure you've checked to ensure these are all that are required for the request to be "proper".

File: src/routes/sunshine.js [hunk 6]
```diff
@@ -94,8 +95,54 @@ module.exports = (app, db) => {
...
+            res.setHeader('Content-Type', 'text/csv');
+            res.write(headersToCSV(csvHeaders));
+            while (await cursor.hasNext()) {
```
Comment (EvilKanoa) [inline src/routes/sunshine.js:132]: Did you fix the issue where it was only writing out the first few entries and not all of them?

File: src/routes/sunshine.js [hunk 7]
```diff
@@ -94,8 +95,54 @@ module.exports = (app, db) => {
...
+            for (header of headers) {
+                line += '"' + header.toString() + '",';
+            }
+            line += '\n';
+            return line;
+        }
+
+        // entry - an object from database to convert to a CSV line
+        // headers - array of headers
+        const entryToCSV = (entry, headers) => {
```
Comment (EvilKanoa) [inline src/routes/sunshine.js:112]: Same as above, this function could be reduced quite a bit using array functions:
```JavaScript
const entryToCSV = (entry, headers) => headers
    .map((header) => entry[header] ? `"${entry[header].toString()}"` : '""')
    .join(',') + '\n';
```

**Review by EvilKanoa — CHANGES_REQUESTED** [review]
Looks solid. Added some suggestions to reduce the LOC a bunch just by using some newer JS features.

File: src/routes/sunshine.js (hunk 6 above)
Comment (jayellul) [reply src/routes/sunshine.js:132]: yes i did. that specific test query wasn't working for some reason but I tried a bunch of other ones and they all worked

File: src/routes/sunshine.js (hunk 3 above)
Comment (jayellul) [reply src/routes/sunshine.js]: Fixed.

File: out.csv (hunk 2 above)
Comment (jayellul) [reply out.csv]: Removed file and pushed.

**Review by EvilKanoa — APPROVED** [review]

**Pull request merged**